### PR TITLE
Add `physical_tile_size` field to determine how large each tile is rendered in-world

### LINKED ISSUE #337: TilemapBundle::grid_size should scale images
I believe it should be in TilemapBundle, but I may be wrong.

Currently, `TilemapBundle::grid_size` cuts images off if smaller than `TilemapBundle::tile_size` or adds padding to the left and above if larger. The expected result is a scaled tile image which matches `TilemapBundle::grid_size`.

This would allow a custom orthographic projection mapping where world space coordinates are the same as tile position without needing to re-adjust the orthographic projection's scale per tileset size. This would also allow use of multiple tileset's where the aspect ratio is identical even if the tile sizes are not. Traditionally Bevy's Sprite::custom_size could be used to achieve the same result.

If I have misinterpreted grid_size, or a breaking change like this is not allowed, an additional field for `TilemapBundle::tile_custom_size` or the like would be requested.

### PR DESCRIPTION
**Pull request by SirHall**
As `bevy_ecs_tilemap` stands now, each tile's size in-world is tied to the pixel-size of the tile in the source atlas.
If I however want my tiles to be of unit size or any particular scale that I require, this is not currently possible.
With this pull request, a user can now set `physical_tile_size` to determine how large a tile should be rendered in-world.

This should also close #337.

### CODE REVIEW

**Review by bzm3r — CHANGES_REQUESTED** [review]
I am not sure about the name `physical_tile_size`. `physical` is a bit of an overloaded term.

I think `TilemapTileSize` should behave exactly like the new `TilemapPhysicalTileSize`.

Instead, there should be a separate type called `SourceTileSize`, which specifies how large a tile is in the source texture. It should not be a field on `TilemapBundle`, instead the variants of `TilemapTexture` should be structs which have a field `tile_size: SourceTileSize`

`TilemapTileSize` should be an option on `TilemapBundle`. Currently we have:

https://github.com/StarArawn/bevy_ecs_tilemap/blob/cac7598fbcd518f9f6c6587add280be30f66d484/src/lib.rs#L103

Instead this should be `pub tile_size: Option<TilemapTileSize>`

If the user wants to specify an explicit `TilemapTileSize`, they can do so by providing `Some(TilemapTileSize { ... })`, otherwise it should be initialized based on the `SourceTileSize` provided for the `TilemapTexture` 

The shaders should get two pieces of info: the `SourceTileSize`, and the `TilemapTileSize`

Let me know if this makes sense to you? I think implementing this shouldn't be too hard based on what you've written so far; it will require some variable renaming and some work on `TilemapTexture`, but will be a clearer solution overall?

@StarArawn @rparrett Let me know if the above review suggestion makes sense to you guys too.

File: examples/bench.rs [hunk 1]
```diff
@@ -25,6 +25,8 @@ fn startup(mut commands: Commands, asset_server: Res<AssetServer>) {
 
     let tile_size = TilemapTileSize { x: 16.0, y: 16.0 };
     let grid_size = tile_size.into();
+
+    println!("{:#?}", grid_size);
```
[the remark this answers is not in the document]
Comment (SirHall) [reply examples/bench.rs:29]: I thought I checked for things like that!
Sorry will remove.

**Comment by SirHall** [comment]
Yeah that's fine, to me 'physical' meant how large it would *'physically'* appear in the game-world.
I'm now thinking it may actually be even clearer to have one be `TilemapSourceTileSize` and the in-world one be `TilemapPhysicalTileSize` to clearly separate both - though I'm fine with any reasonable naming scheme.
I don't have time immediately to go over it so I'll probably have to do it the day after tomorrow for me.

**Comment by bzm3r** [comment]
I would prefer `TilemapSourceTileSize` and `TilemapTileSize`. Or `TextureTileSize` and `TilemapTileSize`. 

I am just not a fan of "physical" because it comes up a lot in gpu stuff with fairly different meanings. In programming more generally, the term "physical" is often contrasted with "logical"; if there is a `TilemapPhysicalTileSize`, is there a `TilemapLogicalTileSize`?

**Comment by rparrett** [comment]
Agree that `TilemapPhysicalTileSize` strikes me as referring to physical pixels on a screen.

> Instead this should be `pub tile_size: Option<TilemapTileSize>`

I don't think bevy will let us do that.

**Comment by SirHall** [comment]
I do think the comparison to GPU architecture is being a little pedantic - but I'm fine with any reasonable option, do let me know with the two names once it's settled.

**Comment by juzi5201314** [comment]
Useful feature, can we still hope to get it? Or can such a function be achieved on existing code?

**Comment by SirHall** [comment]
> Useful feature, can we still hope to get it? Or can such a function be achieved on existing code?

I have been quite busy for a while recently and only a few days thought of updating this PR to finally get it merged in, ideally I can get this merged in shortly.

**Comment by MeoMix** [comment]
just FYI I am working on addressing the PRs concerns in my own branch, should have it done tmmw

**Comment by SirHall** [comment]
Sorry for the long hiatus, this has been reworked, `physical` renamed to `in_world` (as I feel this causes the least confusion) and should be good to merge.

**Comment by SirHall** [comment]
I don't know if @bzm3r is still active, could we have a new reviewer for this PR?

**Comment by SirHall** [comment]
> @SirHall I'm no longer active, sorry! I also apologize if I was too pedantic early on regarding how things ought to be named, and thus delayed getting this merged sooner.

All good, hope you're doing well!

**Pull request closed without merging**